### PR TITLE
Reimplement native methods for reading and storing opcode parameters

### PR DESCRIPTION
**Pull request by x87**

### CODE REVIEW

File: source/CGameVersionManager.cpp [hunk 1]
```diff
@@ -28,6 +28,7 @@ namespace CLEO
         { 0x00465AA0,	memory_und, 0x00465AA0, 0x00465B20, 0x0046B2C0 },		// MA_STOP_SCRIPT_FUNCTION,
         { 0x00465E60,	memory_und, 0x00465E60, 0x00465EE0, 0x0046B640 },		// MA_SCRIPT_OPCODE_HANDLER0_FUNCTION,
         { 0x00464080,	memory_und, 0x00464080, 0x00464100, 0x00469790 },		// MA_GET_SCRIPT_PARAMS_FUNCTION,
+        { 0x00464250,	memory_und, memory_und, memory_und, memory_und },		// MA_GET_NEXT_SCRIPT_PARAM_NO_UPDATE_FUNCTION,
```
Comment (MiranDMC) [inline source/CGameVersionManager.cpp:31]: So it is time to add check for US 1.0 version and throw error message for others.

File: source/CleoBase.cpp [hunk 2]
```diff
@@ -6,6 +6,7 @@ namespace CLEO
 {
     CCleoInstance CleoInstance;
     CCleoInstance& GetInstance() { return CleoInstance; }
+    std::stack<int>         CleoStack;
```
Comment (MiranDMC) [inline source/CleoBase.cpp:9]: Seems like it should hold SCRIPT_VAR type. Also what with the strings?

Comment (x87) [reply source/CleoBase.cpp:9]: strings are out of scope for now

File: cleo_sdk/CLEO.h [hunk 3]
```diff
@@ -69,6 +69,8 @@ enum eDataType : BYTE
 	DT_LVAR_STRING, // localVarVString @v
 	DT_VAR_STRING_ARRAY, // globalVarStringArr v$(,)
 	DT_LVAR_STRING_ARRAY, // localVarStringArr @v(,)
+	DT_PUSH,
```
Comment (MiranDMC) [inline cleo_sdk/CLEO.h:72]: add comments

File: cleo_sdk/CLEO.h [hunk 4]
```diff
@@ -78,6 +80,7 @@ enum eArrayDataType : BYTE
 	ADT_FLOAT, // variable with integer
 	ADT_TEXTLABEL, // variable with short string (8 char)
 	ADT_STRING, // variable with long string (16 char)
+	ADT_POP,
```
Comment (MiranDMC) [inline cleo_sdk/CLEO.h:83]: add comment

File: source/CScriptEngine.cpp [hunk 5]
```diff
@@ -347,6 +348,289 @@ namespace CLEO
         GetInstance().ScriptEngine.ReregisterAllScripts();
     }
 
+    SCRIPT_VAR* CRunningScript::GetPointerToLocalVariable(CRunningScript* script, int varIndex) {
```
Comment (MiranDMC) [inline source/CScriptEngine.cpp:351]: Why this is method of CRunningScript and still takes script as input argument?

File: source/CScriptEngine.cpp [hunk 6]
```diff
@@ -347,6 +348,289 @@ namespace CLEO
         GetInstance().ScriptEngine.ReregisterAllScripts();
     }
 
+    SCRIPT_VAR* CRunningScript::GetPointerToLocalVariable(CRunningScript* script, int varIndex) {
+        if (IsMission())
```
Comment (MiranDMC) [inline source/CScriptEngine.cpp:352]: script->IsMission() ?

File: source/CScriptEngine.cpp [hunk 7]
```diff
@@ -347,6 +348,289 @@ namespace CLEO
...
+            return script->GetVarPtr(varIndex);
+    }
+
+    void CRunningScript::ReadArrayInformation(CRunningScript* script, WORD* outArrVarOffset, int* outArrElemIdx) {
+        BYTE* ip = script->GetBytePointer();
+
+        *outArrVarOffset = script->ReadDataWord();
+        WORD arrayIndexVar = script->ReadDataWord();
+        script->IncPtr(); // skip size
+		BYTE arrayType = script->ReadDataByte();
```
Comment (MiranDMC) [inline source/CScriptEngine.cpp:364]: There are already types like eArrayDataType and ArrayDataTypeMask in CLEO.h

File: source/CScriptEngine.cpp [hunk 8]
```diff
@@ -347,6 +348,289 @@ namespace CLEO
...
+            *outArrElemIdx = GetPointerToLocalVariable(script, arrayIndexVar)->dwParam;
+        }
+    }
+
+    void CRunningScript::CollectParams(CRunningScript* script, WORD count) {
+        WORD arrVarOffset;
+        int arrElemIdx;
+
+        for (int i = 0; i < count; i++) {
+            int dt = script->ReadDataByte();
```
Comment (MiranDMC) [inline source/CScriptEngine.cpp:386]: Use ReadDataType() instead?

File: source/CScriptEngine.cpp [hunk 9]
```diff
@@ -347,6 +348,289 @@ namespace CLEO
...
+            break;
+        case DT_LVAR_ARRAY:
+            ReadArrayInformation(script, &arrVarOffset, &arrElemIdx);
+            result = *GetPointerToLocalVariable(script, arrVarOffset + arrElemIdx);
+            break;
+        case DT_POP:
+            break;
+        }
+
+        script->SetIp(ip);
```
Comment (MiranDMC) [inline source/CScriptEngine.cpp:544]: When rewinding please make sure `GetInstance().OpcodeSystem.handledParamCount` is reverted to correct state too. It is used for debug messages. Maybe there is way to handle it some smarter way.

File: source/CScriptEngine.cpp [hunk 10]
```diff
@@ -347,6 +348,289 @@ namespace CLEO
...
+        }
+    }
+
+
+    SCRIPT_VAR* CRunningScript::GetPointerToScriptVariable(CRunningScript* script, BYTE type) {
+        BYTE arrElemSize;
+        WORD arrVarOffset;
+        int  arrElemIdx;
+
+        int dt = script->ReadDataByte();
```
Comment (MiranDMC) [inline source/CScriptEngine.cpp:568]: GetDataType

File: source/CleoBase.cpp (hunk 2 above)
Comment (MiranDMC) [inline source/CleoBase.cpp:9]: Global variable? Seems like this should live inside CCustomOpcodeSystem

File: source/CleoBase.h [hunk 11]
```diff
@@ -52,6 +54,9 @@ namespace CLEO
         void CallCallbacks(eCallbackId id);
         void CallCallbacks(eCallbackId id, DWORD arg);
 
+        void StackPush(int val);
```
Comment (MiranDMC) [inline source/CleoBase.h:57]: SCRIPT_VAR as param.

**Pull request closed without merging**